integration-tests/src/tests/client/features/flat_storage.rs

```diff
@@ -0,0 +1,151 @@
+use near_primitives_core::config::ExtCosts;
+use near_primitives_core::hash::CryptoHash;
+use near_primitives_core::types::Gas;
+use near_store::test_utils::create_test_store;
+use nearcore::config::GenesisExt;
+use nearcore::TrackedConfig;
+use std::path::Path;
+use std::sync::Arc;
+
+/// Check that after flat storage upgrade:
+/// - value read from contract is the same;
+/// - touching trie node cost for read decreases to zero.
+#[test]
+fn test_flat_storage_upgrade() {
+    let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
+    let epoch_length = 12;
+    let old_protocol_version = ProtocolFeature::FlatStorageReads.protocol_version() - 1;
+    let new_protocol_version = old_protocol_version + 1;
+    genesis.config.epoch_length = epoch_length;
+    genesis.config.protocol_version = old_protocol_version;
+    let chain_genesis = ChainGenesis::new(&genesis);
+    let runtimes: Vec<Arc<dyn RuntimeWithEpochManagerAdapter>> =
+        vec![Arc::new(nearcore::NightshadeRuntime::test_with_runtime_config_store(
+            Path::new("../../../.."),
+            create_test_store(),
+            &genesis,
+            TrackedConfig::new_empty(),
+            RuntimeConfigStore::new(None),
+        ))];
+    let mut env = TestEnv::builder(chain_genesis).runtime_adapters(runtimes).build();
+
+    // Deploy contract to state.
+    deploy_test_contract_with_protocol_version(
+        &mut env,
+        "test0".parse().unwrap(),
+        near_test_contracts::base_rs_contract(),
+        epoch_length / 3,
+        1,
+        old_protocol_version,
+    );
+
+    let signer = InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+    let gas = 20_000_000_000_000;
+    let tx = Transaction {
+        signer_id: "test0".parse().unwrap(),
+        receiver_id: "test0".parse().unwrap(),
+        public_key: signer.public_key(),
+        actions: vec![],
+        nonce: 0,
+        block_hash: CryptoHash::default(),
+    };
+
+    // Write key-value pair to state.
+    {
+        let write_value_action = vec![Action::FunctionCall(FunctionCallAction {
+            args: encode(&[1u64, 10u64]),
+            method_name: "write_key_value".to_string(),
+            gas,
+            deposit: 0,
+        })];
+        let tip = env.clients[0].chain.head().unwrap();
+        let signed_transaction = Transaction {
+            nonce: 10,
+            block_hash: tip.last_block_hash,
+            actions: write_value_action,
+            ..tx.clone()
+        }
+        .sign(&signer);
+        let tx_hash = signed_transaction.get_hash();
+        env.clients[0].process_tx(signed_transaction, false, false);
+        for i in 0..epoch_length / 3 {
+            env.produce_block(0, tip.height + i + 1);
+        }
+
+        let final_result = env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap();
+        assert_matches!(final_result.status, FinalExecutionStatus::SuccessValue(_));
```

Comment:
I think this could reuse

https://github.com/near/nearcore/blob/229cf47a9c7bc9ae90cbf69f5d10bb00983fb6ab/core/primitives/src/test_utils.rs#L519-L530

which avoid duplication and also gives better error messages when the test fails due to a receipt.

```suggestion
        let final_result = env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap().assert_success();
```